File: web/availability.js
/*
 * RoomRadar 空き判定（クライアント側）.
 *
 * roomradar/availability.py・terms.py を JavaScript へ移植した純粋関数群。
 * ブラウザ（<script> で window.RoomRadar）でも Node（require）でも動く UMD。
 * DESIGN.md §4.1「静的コア」：ビルド済み JSON を読んでブラウザ側で空きを計算する。
 */
(function (root, factory) {
  if (typeof module === "object" && module.exports) module.exports = factory();
  else root.RoomRadar = factory();
})(typeof self !== "undefined" ? self : this, function () {
  "use strict";

  function parseMD(s) {
    const parts = String(s).split("-");
    return [Number(parts[0]), Number(parts[1])];
  }
  function cmpMD(a, b) {
    return a[0] - b[0] || a[1] - b[1];
  }

  // today が start〜end（両端含む）に入るか。start>end は年末跨ぎ。
  function inSeason(today, start, end) {
    if (cmpMD(start, end) <= 0) return cmpMD(start, today) <= 0 && cmpMD(today, end) <= 0;
    return cmpMD(today, start) >= 0 || cmpMD(today, end) <= 0;
  }

  // 有効な学期 id の一覧（config 定義順）。
  function activeTerms(terms, date) {
    date = date || new Date();
    const today = [date.getMonth() + 1, date.getDate()];
    const out = [];
    for (const t of terms || []) {
      if (t.always) out.push(t.id);
      else if (t.start && t.end && inSeason(today, parseMD(t.start), parseMD(t.end))) out.push(t.id);
    }
    return out;
  }

  // 指定 曜日×時限 の空き教室。Python 版 free_rooms と同じ結果・同じ並び。
  function freeRooms(data, opts) {
    opts = opts || {};
    const day = opts.day;
    const period = String(opts.period);
    const building = opts.building || null;
    const today = opts.today || new Date();

    const active = new Set(activeTerms(data.terms, today));
    const occupied = new Set();
    const occ = data.occupied || {};
    for (const term of Object.keys(occ)) {
      if (!active.has(term)) continue;
      const rooms = (((occ[term] || {})[day] || {})[period]) || [];
      for (const r of rooms) occupied.add(r);
    }

    const order = new Map((data.buildings || []).map((b, i) => [b.name, i]));
    const fallback = order.size;
    const cand = (data.rooms || []).filter(
      (r) => !occupied.has(r.room) && (!building || r.building === building)
    );
    cand.sort((a, b) => {
      const oa = order.has(a.building) ? order.get(a.building) : fallback;
      const ob = order.has(b.building) ? order.get(b.building) : fallback;
      if (oa !== ob) return oa - ob;
      return a.room < b.room ? -1 : a.room > b.room ? 1 : 0;
    });
    return cand;
  }

  // 現在の曜日・時限（検索の初期値）。
  function currentDayPeriod(data, now) {
    now = now || new Date();
    const names = ["月", "火", "水", "木", "金", "土", "日"]; // 0=月 … 6=日
    let day = names[(now.getDay() + 6) % 7]; // JS: 0=日
    if (!data.days.includes(day)) day = data.days[0];
    const hhmm =
      String(now.getHours()).padStart(2, "0") + ":" + String(now.getMinutes()).padStart(2, "0");
    let period = data.periods[0].period;
    for (const p of data.periods) {
      if (p.start <= hhmm && hhmm <= p.end) {
        period = p.period;
        break;
      }
    }
    return { day: day, period: period };
  }

  return {
    activeTerms: activeTerms,
    inSeason: inSeason,
    freeRooms: freeRooms,
    currentDayPeriod: currentDayPeriod,
  };
});

File: roomradar/availability.py
"""空き判定エンジン（学校非依存の純粋関数）.

現行 ``nust-room-search`` の ``index()`` 内ロジック
（指定 曜日×時限×有効学期 の使用中教室集合を作り、全教室から差集合を取る）を、
校舎別 ``if`` 分岐を撤去して学校非依存に一般化したもの（DESIGN.md §8-2）。
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Sequence


@dataclass(frozen=True)
class Lesson:
    """時間割の 1 コマ（``schedule.csv`` の 1 行）."""

    term: str
    day: str
    period: int
    room: str
    building: str
    department: str = ""
    course: str = ""


@dataclass(frozen=True)
class Room:
    """教室マスタの 1 室（``classrooms.csv`` の 1 行）."""

    room: str
    building: str


def occupied_rooms(
    lessons: Iterable[Lesson], *, day: str, period: int, active_terms: Sequence[str]
) -> set[str]:
    """指定 曜日×時限 に、有効学期のいずれかで授業が入っている教室名の集合."""
    active = set(active_terms)
    return {
        lesson.room
        for lesson in lessons
        if lesson.day == day and lesson.period == period and lesson.term in active
    }


def free_rooms(
    lessons: Iterable[Lesson],
    rooms: Iterable[Room],
    *,
    day: str,
    period: int,
    active_terms: Sequence[str],
    building_order: Sequence[str] | None = None,
    building: str | None = None,
) -> list[Room]:
    """指定 曜日×時限 に空いている教室を返す.

    Args:
        lessons: 学校の時間割（全コマ）。
        rooms: 学校の全教室（母集合）。
        day: 曜日（config の ``days`` のいずれか）。
        period: 時限（config の ``periods[].no`` のいずれか）。
        active_terms: 有効な学期 id の一覧（:func:`roomradar.terms.active_terms`）。
        building_order: 表示・ソート順に使う校舎名の並び（config の定義順）。
        building: 指定すると、その校舎名の教室だけに絞り込む。

    Returns:
        空き教室の一覧。``building_order`` → 教室名 の順でソート済み。
    """
    occupied = occupied_rooms(lessons, day=day, period=period, active_terms=active_terms)
    order = {name: i for i, name in enumerate(building_order or [])}
    fallback = len(order)  # 並びに無い校舎は末尾へ

    candidates = [
        room
        for room in rooms
        if room.room not in occupied and (building is None or room.building == building)
    ]
    candidates.sort(key=lambda r: (order.get(r.building, fallback), r.room))
    return candidates
